mysql/mysqldump.go

```diff
@@ -410,9 +408,26 @@ func getTypeModsAndID(conv *internal.Conv, columnType string) (string, []int64)
 
 // handleParseError handles error while parsing mysqldump
 // statements and attempts at creating parsable chunk.
-// Error can be due to insert statement or unsupported Spatial
-// datatypes in create statement.
+// Error can be due to insert statement, unsupported Spatial
+// datatypes in create statement or unsupported stored programs.
 func handleParseError(conv *internal.Conv, chunk string, err error, l [][]byte) ([]ast.StmtNode, bool) {
+	// Check error for statements that are not supported by Pingcap parser.
+	// It includes delimiter, function, procedures and triggers.
+	// In case it's due to 'DELIMITER', we raparse till the chunk
+	// contains 2 delimiters. Eg:
+	//  1) 'DELIMITER $$' - First one is to define a delimiter
+	//     for functions, stored procedures or triggers.
+	//  2) 'DELIMITER ;' - It is restored to default delimiter i.e. ';'
+	// OR,
+	// Case when it's just due to functions, procedures or triggers
+	// without a delimiter statement.
+	errMsg := strings.ToLower(err.Error())
+	if unsupportedRegexp.MatchString(errMsg) || strings.Contains(errMsg, "delimiter") {
+		if strings.Count(strings.ToLower(chunk), "delimiter") == 1 {
```

Comment:
So if there was only one delimiter statement, we'd skip all statements from the delimiter statement onwards to the end of the file, at which point we'll generate a "Error parsing last NNN lines of input".

I don't have any good suggestions for handling this better.

If I understand correctly, mysqldump doesn't generate such dump files: it only generates dump files with pairs of delimiter statements -- is that a reasonable to assume?

Reply:
Yes correct, MySQL generates dump with pairs of delimiter only and is limited to functions, stored procedures and triggers. I have also confirmed this with our DBA team.

> So if there was only one delimiter statement, we'd skip all statements from the delimiter statement onwards to the end of the file, at which point we'll generate a "Error parsing last NNN lines of input".

If dump file is generated using `mysqldump` command or best practices are followed as recommended by MySQL, then this case is not possible. But if anyone has written a dump file manually then it could be possible. 

Overall it's fair to assume that the dump file will be as per the standards.

Reply:
Thanks!